### PR TITLE
implemented gpuarray.stack(),gpuarray.concatenate()

### PR DESCRIPTION
**Pull request by mitkotak**

### CODE REVIEW

File: pycuda/gpuarray.py [hunk 1]
```diff
@@ -1768,6 +1768,113 @@ def make_func_for_chunk_size(chunk_size):
 # {{{ shape manipulation
 
 
+def concatenate(arrays, axis=0, queue=None, allocator=None):
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: docs

File: pycuda/gpuarray.py [hunk 2]
```diff
@@ -1768,6 +1768,113 @@ def make_func_for_chunk_size(chunk_size):
...
+
+    # {{{ find properties of result array
+
+    shape = None
+
+    for i_ary, ary in enumerate(arrays):
+        queue = queue or ary.queue
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: remove the queue

File: pycuda/gpuarray.py [hunk 3]
```diff
@@ -1768,6 +1768,113 @@ def make_func_for_chunk_size(chunk_size):
...
+                        "(shold have %d, has %d)"
+                        % (i_ary, len(ary.shape), len(shape)))
+
+                ary_shape_list = list(ary.shape)
+                if (ary_shape_list[:axis] != shape[:axis]
+                        or ary_shape_list[axis+1:] != shape[axis+1:]):
+                    raise ValueError("%d'th array has residual not matching "
+                            "other arrays" % i_ary)
+
+                shape[axis] += ary.shape[axis]
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: Avoid stateful initializations. Possibly what pytato does: https://github.com/inducer/pytato/blob/8f6a07fe041bc5a4bf35e770005f1b25f9b2efb0/pytato/array.py#L1673-L1680

File: pycuda/gpuarray.py [hunk 4]
```diff
@@ -1768,6 +1768,113 @@ def make_func_for_chunk_size(chunk_size):
...
+
+    # }}}
+
+    shape = tuple(shape)
+    dtype = np.find_common_type([ary.dtype for ary in arrays], [])
+    result = empty(queue, shape, dtype, allocator=allocator)
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: Looks like `pycuda.gpuarray` doesn't have an empty. Make one?
Implementation almost provided: https://github.com/inducer/pycuda/blob/f65bdbdba5a33652933557412ba153f673977361/pycuda/gpuarray.py#L1277-L1281

File: pycuda/gpuarray.py [hunk 5]
```diff
@@ -1768,6 +1768,113 @@ def make_func_for_chunk_size(chunk_size):
 # {{{ shape manipulation
 
 
+def concatenate(arrays, axis=0, queue=None, allocator=None):
+
+    # {{{ find properties of result array
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py:1754]: Add a comment somewhere in the function body that "impllementation is borrowed from pyopencl.array`.

File: pycuda/gpuarray.py [hunk 6]
```diff
@@ -1768,6 +1768,113 @@ def make_func_for_chunk_size(chunk_size):
...
+    """
+
+    if not arrays:
+        raise ValueError("need at least one array to stack")
+
+    input_shape = arrays[0].shape
+    input_ndim = arrays[0].ndim
+    axis = input_ndim if axis == -1 else axis
+
+    if queue is None:
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: queue?

File: pycuda/gpuarray.py [hunk 7]
```diff
@@ -1768,6 +1768,113 @@ def make_func_for_chunk_size(chunk_size):
...
+
+    if (axis == input_ndim and not all(ary.flags.f_contiguous
+                                               for ary in arrays)):
+        # pycuda.Array.__setitem__ does not support non-contiguous assignments
+        raise NotImplementedError
+
+
+    result_shape = input_shape[:axis] + (len(arrays),) + input_shape[axis:]
+
+    result = empty(queue, result_shape, np.result_type(*(ary.dtype
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: empty (see above)

File: test/test_gpuarray.py [hunk 8]
```diff
@@ -468,6 +468,46 @@ def test_arange(self):
...
+        shape = (2, 2, 2)[:input_dims]
+        axis = -1 if order == "F" else 0
+
+        from numpy.random import default_rng
+        rng = default_rng()
+        x_in = rng.random(size=shape)
+        y_in = rng.random(size=shape)
+        x_in = x_in if order == "C" else np.asfortranarray(x_in)
+        y_in = y_in if order == "C" else np.asfortranarray(y_in)
+
```
Comment (kaushikcfd) [inline test/test_gpuarray.py:481]: Doesn't actually test anything.

**Review by kaushikcfd — CHANGES_REQUESTED** [review]
Some comments to begin with.

File: pycuda/gpuarray.py (hunk 1 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: added docs and removed queue

File: pycuda/gpuarray.py (hunk 2 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: queue removed

File: pycuda/gpuarray.py (hunk 3 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: added state initializations

File: pycuda/gpuarray.py (hunk 5 above)
Comment (mitkotak) [reply pycuda/gpuarray.py:1754]: created empty

File: pycuda/gpuarray.py (hunk 6 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: removed queue

File: pycuda/gpuarray.py (hunk 7 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: added empty

File: test/test_gpuarray.py (hunk 8 above)
Comment (mitkotak) [reply test/test_gpuarray.py:481]: test implemented

File: pycuda/gpuarray.py (hunk 1 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py]: I don't see the commit for it. Could you push your changes.

**Comment by mitkotak** [comment]
Yup I was still debugging them but let me push the changes I guess

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

File: doc/array.rst [hunk 9]
```diff
@@ -339,6 +339,15 @@ Constructing :class:`GPUArray` Instances
     Return the :class:`GPUArray` ``[a[indices[0]], ..., a[indices[n]]]``.
     For the moment, *a* must be a type that can be bound to a texture.
 
+.. function:: concatenate(arrays, axis=0, allocator=None)
+
+   Join a sequence of arrays along an existing axis.
+
+.. function:: stack(arrays, axis=0, allocator=None)
+
+   Joing a sequeunce of arrays along a new axis
```
Comment (kaushikcfd) [inline doc/array.rst]: - Typo
- Period at the end.

File: pycuda/gpuarray.py [hunk 10]
```diff
@@ -1767,6 +1767,104 @@ def make_func_for_chunk_size(chunk_size):
 
 # {{{ shape manipulation
 
+def concatenate(arrays, axis=0, allocator=None):
+
+    # impllementation is borrowed from pyopencl.array.concatenate()
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: - Typo.
- Should come after docstring.

File: pycuda/gpuarray.py [hunk 11]
```diff
@@ -1767,6 +1767,104 @@ def make_func_for_chunk_size(chunk_size):
...
+    # {{{ find properties of result array
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: Move the [fold marker](https://vim.fandom.com/wiki/Folding#Folding_with_markers) after the docstring.

File: pycuda/gpuarray.py [hunk 12]
```diff
@@ -1767,6 +1767,104 @@ def make_func_for_chunk_size(chunk_size):
...
+        my_len = ary.shape[axis]
+        result[full_slice[:axis] + (slice(base_idx, base_idx+my_len),) + full_slice[axis+1:]] = ary
+        base_idx += my_len
+
+    return result
+
+
+def stack(arrays, axis=0, allocator=None):
+    """
+    implementation is borrowed from pyopencl.array.stack()
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: "implementation is borrowed from ...." shouldn't end up in the docstring.

File: test/test_gpuarray.py [hunk 13]
```diff
@@ -468,6 +468,56 @@ def test_arange(self):
...
+        np.testing.assert_allclose(gpuarray.stack((x_gpu, y_gpu), axis=axis).get(),
+                                np.stack((x_in, y_in), axis=axis))
+        
+        x_gpu_shape = x_gpu.shape
+        y_gpu_shape = y_gpu.shape
+
+        np.testing.assert_allclose(x_gpu_shape,shape)
+
+        np.testing.assert_allclose(y_gpu_shape,shape)
```
Comment (kaushikcfd) [inline test/test_gpuarray.py]: What is this supposed to check? `x_in_gpu` is initialized to be an array of shape `shape` and that is again being checked here. I.e. this test is trivial. You probably want to check the equality of shapes of the gpuarray stack result and the numpy array result.

File: test/test_gpuarray.py [hunk 14]
```diff
@@ -468,6 +468,56 @@ def test_arange(self):
...
+        b_dev = curand((4, 15, 20), dtype=np.float32)
+        c_dev = curand((3, 15, 20), dtype=np.float32)
+        a = a_dev.get()
+        b = b_dev.get()
+        c = c_dev.get()
+
+        cat_dev = gpuarray.concatenate((a_dev, b_dev, c_dev))
+        cat = np.concatenate((a, b, c))
+
+        assert la.norm(cat - cat_dev.get()) == 0
```
Comment (kaushikcfd) [inline test/test_gpuarray.py]: Equality for floating point numbers is dangerous. Either use some tolerance or just `np.testing.assert_allclose`.

File: pycuda/gpuarray.py [hunk 15]
```diff
@@ -1767,6 +1767,104 @@ def make_func_for_chunk_size(chunk_size):
...
+
+    if (axis == 0 and not all(ary.flags.c_contiguous
+                                       for ary in arrays)):
+        # pycuda.GPUArray.__setitem__ does not support non-contiguous assignments
+        raise NotImplementedError
+
+    if (axis == input_ndim and not all(ary.flags.f_contiguous
+                                               for ary in arrays)):
+        # pycuda.Array.__setitem__ does not support non-contiguous assignments
+        raise NotImplementedError
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: Are you sure pycuda arrays do not support non-contiguous array assignment?
```python
In [9]: import numpy as np

In [10]: import pycuda.gpuarray as gpuarray

In [11]: import pycuda.autoinit

In [12]: a_gpu = gpuarray.zeros((10, ), dtype=np.float32)

In [13]: a_gpu[::2] = 42*np.ones(5, dtype=np.float32)

In [14]: a_gpu
Out[14]: array([42.,  0., 42.,  0., 42.,  0., 42.,  0., 42.,  0.], dtype=float32)
```

File: test/test_gpuarray.py [hunk 16]
```diff
@@ -468,6 +468,56 @@ def test_arange(self):
...
+
+# }}}
+
+# {{{ conditionals, any, all
+    
+
```
Comment (kaushikcfd) [inline test/test_gpuarray.py]: Spurious diff, shouldn't be a part of this patch.

**Review by kaushikcfd — COMMENTED** [review]
Thanks for working on this! Left some more comments.

File: test/test_gpuarray.py (hunk 8 above)
Comment (kaushikcfd) [reply test/test_gpuarray.py:481]: Can be resolved.

File: pycuda/gpuarray.py (hunk 7 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py]: Thanks, can be resolved.

File: pycuda/gpuarray.py (hunk 6 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py]: Thanks, can be resolved.

File: pycuda/gpuarray.py (hunk 5 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py:1754]: Thanks, can be resolved.

File: pycuda/gpuarray.py (hunk 4 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py]: Can be resolved.

File: pycuda/gpuarray.py (hunk 3 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py]: Can be resolved.

File: pycuda/gpuarray.py (hunk 2 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py]: Thanks, can be resolved.

File: pycuda/gpuarray.py (hunk 1 above)
Comment (kaushikcfd) [reply pycuda/gpuarray.py]: Thanks, can be resolved.

File: doc/array.rst (hunk 9 above)
Comment (mitkotak) [reply doc/array.rst]: Thanks for finding the typos. The typo has been fixed in my latest commit for array.rst

**Review by mitkotak — COMMENTED** [review]
Fixed the typo

File: pycuda/gpuarray.py (hunk 10 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: Thanks for finding the typo. The typo has been fixed and the acknowledgment as well fold marker have been relocated.

File: pycuda/gpuarray.py (hunk 11 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: ^

File: pycuda/gpuarray.py (hunk 12 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: Thanks for correcting me there. The acknowledgment has been relocated.

File: pycuda/gpuarray.py (hunk 15 above)
Comment (mitkotak) [reply pycuda/gpuarray.py]: Ohh got it. In that case let me get rid of those 2 checks then. Thanks for the quick demo

File: test/test_gpuarray.py (hunk 14 above)
Comment (mitkotak) [reply test/test_gpuarray.py]: Thanks for finding the mistake. Yup will use assert_allclose instead

File: test/test_gpuarray.py (hunk 16 above)
Comment (mitkotak) [reply test/test_gpuarray.py]: Yup. Will get rid of it then

File: test/test_gpuarray.py (hunk 13 above)
Comment (mitkotak) [reply test/test_gpuarray.py]: Gotcha !! Thanks for the detailed explanation

File: pycuda/gpuarray.py [hunk 17]
```diff
@@ -1767,6 +1767,93 @@ def make_func_for_chunk_size(chunk_size):
...
+        allocator = allocator or ary.allocator
+
+        if shape is None:
+            # first array
+            shape = list(ary.shape)
+
+        else:
+            if len(ary.shape) != len(shape):
+                raise ValueError("%d'th array has different number of axes "
+                        "(shold have %d, has %d)"
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]:
```suggestion
                        "(should have %d, has %d)"
```

File: pycuda/gpuarray.py [hunk 18]
```diff
@@ -1767,6 +1767,93 @@ def make_func_for_chunk_size(chunk_size):
...
+    axis = input_ndim if axis == -1 else axis
+
+    if not all(ary.shape == input_shape for ary in arrays[1:]):
+        raise ValueError("arrays must have the same shape")
+
+    if not (0 <= axis <= input_ndim):
+        raise ValueError("invalid axis")
+
+    result_shape = input_shape[:axis] + (len(arrays),) + input_shape[axis:]
+    result = empty(shape=result_shape, dtype=np.result_type(*(ary.dtype for ary in arrays)), allocator=allocator, order="C" if axis == 0 else "F")
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py]: Trim lines to 84 characters.

File: test/test_gpuarray.py [hunk 19]
```diff
@@ -468,6 +468,49 @@ def test_arange(self):
         a = gpuarray.arange(12, dtype=np.float32)
         assert (np.arange(12, dtype=np.float32) == a.get()).all()
 
+    @mark_cuda_test
+    def test_stack(self):
+
+        orders = ["F", "C"]
+        input_dims = 2
```
Comment (kaushikcfd) [inline test/test_gpuarray.py]: Vary the input_dims here as well.

File: test/test_gpuarray.py [hunk 20]
```diff
@@ -468,6 +468,49 @@ def test_arange(self):
...
+            x_in = x_in if order == "C" else np.asfortranarray(x_in)
+            y_in = y_in if order == "C" else np.asfortranarray(y_in)
+
+            x_gpu = gpuarray.to_gpu(x_in)
+            y_gpu = gpuarray.to_gpu(y_in)
+
+            np.testing.assert_allclose(gpuarray.stack((x_gpu, y_gpu), axis=axis).get(),
+                    np.stack((x_in, y_in), axis=axis))
+
+            np.testing.assert_allclose(gpuarray.stack((x_gpu, y_gpu), axis=axis).shape,
```
Comment (kaushikcfd) [inline test/test_gpuarray.py]: - Don't recompute the stack results
- For the shape checking just use `assert` no need to bring in `assert_allclose`.

File: test/test_gpuarray.py [hunk 21]
```diff
@@ -468,6 +468,49 @@ def test_arange(self):
...
+        b_dev = curand((4, 15, 20), dtype=np.float32)
+        c_dev = curand((3, 15, 20), dtype=np.float32)
+        a = a_dev.get()
+        b = b_dev.get()
+        c = c_dev.get()
+
+        cat_dev = gpuarray.concatenate((a_dev, b_dev, c_dev))
+        cat = np.concatenate((a, b, c))
+
+        np.testing.assert_allclose(cat, cat_dev.get())
```
Comment (kaushikcfd) [inline test/test_gpuarray.py]: Test the shapes here too.

**Review by kaushikcfd — APPROVED** [review]
LGTM, except those minor things.

File: test/test_gpuarray.py [hunk 22]
```diff
@@ -468,6 +468,53 @@ def test_arange(self):
...
+                x_in = x_in if order == "C" else np.asfortranarray(x_in)
+                y_in = y_in if order == "C" else np.asfortranarray(y_in)
+
+                x_gpu = gpuarray.to_gpu(x_in)
+                y_gpu = gpuarray.to_gpu(y_in)
+
+                numpy_stack = np.stack((x_in, y_in), axis=axis)
+                gpuarray_stack = gpuarray.stack((x_gpu, y_gpu), axis=axis)
+
+                np.testing.assert_equal(gpuarray_stack.get(), numpy_stack)
```
Comment (kaushikcfd) [inline test/test_gpuarray.py:495]:
```suggestion
                np.testing.assert_allclose(gpuarray_stack.get(), numpy_stack)
```
Checking equality of floating point arrays is almost never a good idea.

File: test/test_gpuarray.py [hunk 23]
```diff
@@ -468,6 +468,53 @@ def test_arange(self):
...
+
+                np.testing.assert_equal(gpuarray_stack.shape, numpy_stack.shape)
```
Comment (kaushikcfd) [inline test/test_gpuarray.py:497]:
```suggestion
                assert gpuarray_stack.shape == numpy_stack.shape
```

File: test/test_gpuarray.py [hunk 24]
```diff
@@ -468,6 +468,53 @@ def test_arange(self):
...
+        b_dev = curand((4, 15, 20), dtype=np.float32)
+        c_dev = curand((3, 15, 20), dtype=np.float32)
+        a = a_dev.get()
+        b = b_dev.get()
+        c = c_dev.get()
+
+        cat_dev = gpuarray.concatenate((a_dev, b_dev, c_dev))
+        cat = np.concatenate((a, b, c))
+
+        np.testing.assert_equal(cat, cat_dev.get())
```
Comment (kaushikcfd) [inline test/test_gpuarray.py:514]:
```suggestion
        np.testing.assert_allclose(cat, cat_dev.get())
```

File: test/test_gpuarray.py [hunk 25]
```diff
@@ -468,6 +468,53 @@ def test_arange(self):
...
+
+        np.testing.assert_equal(cat.shape, cat_dev.shape)
```
Comment (kaushikcfd) [inline test/test_gpuarray.py:516]:
```suggestion
        assert cat.shape == cat_dev.shape
```

File: pycuda/gpuarray.py [hunk 26]
```diff
@@ -1767,6 +1767,95 @@ def make_func_for_chunk_size(chunk_size):
...
+        else:
+            if len(ary.shape) != len(shape):
+                raise ValueError("%d'th array has different number of axes "
+                        "(should have %d, has %d)"
+                        % (i_ary, len(ary.shape), len(shape)))
+
+            if (ary.ndim != arrays[0].ndim
+                    or shape_except_axis(ary) != shape_except_axis(arrays[0])):
+                raise ValueError("%d'th array has residual not matching "
+                        "other arrays" % i_ary)
```
Comment (kaushikcfd) [inline pycuda/gpuarray.py:1777]:
```suggestion
                raise ValueError(f"{i_ary}'th array has residual not matching "
                                 "other arrays")
```

**Review by kaushikcfd — APPROVED** [review]
Thanks, LGTM. Few minor things.

**Pull request merged**